src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,916 @@
+"""
+@inline function prim2cons(prim, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = prim
+
+  cons = zero(MVector{nvariables(equations), eltype(prim)})
+  cons[1] = B1
+  cons[2] = B2
+  cons[3] = B3
+  for k in eachcomponent(equations)
+    rho, v1, v2, v3, p = get_component(k, prim, equations)
+    rho_v1 = rho * v1
+    rho_v2 = rho * v2
+    rho_v3 = rho * v3
+
+    rho_e = p/(gammas[k] - 1.0) + 0.5 * (rho_v1 * v1 + rho_v2 * v2 + rho_v3 * v3) +
+                                  0.5 * (B1^2 + B2^2 + B3^2)
+    
+    set_component!(cons, k, rho, rho_v1, rho_v2, rho_v3, rho_e, equations)
+  end
+
+  return SVector(cons)
+end
+
+"""
+Compute the fastest wave speed for ideal MHD equations: c_f, the fast magnetoacoustic eigenvalue
+  !!! ATTENTION: This routine is provisional.. Change once the fastest wave speed is known!!
+"""
+@inline function calc_fast_wavespeed(cons, orientation::Integer, equations::IdealMhdMultiIonEquations2D)
+  B1, B2, B3, _ = cons
+
+  c_f = zero(real(equations))
+  for k in eachcomponent(equations)
+    rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, cons, equations)
+    
+    v1 = rho_v1 / rho
+    v2 = rho_v2 / rho
+    v3 = rho_v3 / rho
+    v_mag = sqrt(v1^2 + v2^2 + v3^2)
+    gamma = equations.gammas[k]
+    p = (gamma - 1)*(rho_e - 0.5*rho*v_mag^2 - 0.5*(B1^2 + B2^2 + B3^2))
+    a_square = gamma * p / rho
+    sqrt_rho = sqrt(rho)
+
+    b1 = B1 / sqrt_rho
+    b2 = B2 / sqrt_rho
+    b3 = B3 / sqrt_rho
+    b_square = b1^2 + b2^2 + b3^2
+
+    if orientation == 1
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b1^2)))
+    else #if orientation == 2
+      c_f = max(c_f, sqrt(0.5*(a_square + b_square) + 0.5*sqrt((a_square + b_square)^2 - 4.0*a_square*b2^2)))
+    end
+  end
+
+  return c_f
+end
+
+"""
+Routine to compute the Charge-averaged velocities:
+* v*_plus: Charge-averaged velocity
+* vk*_plus: Contribution of each species to the charge-averaged velocity
+"""
+@inline function charge_averaged_velocities(u, equations::IdealMhdMultiIonEquations2D)
+
+  total_electron_charge = zero(real(equations))
+  
+  vk1_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk2_plus = zero(MVector{ncomponents(equations), eltype(u)})
+  vk3_plus = zero(MVector{ncomponents(equations), eltype(u)})
+
+  for k in eachcomponent(equations)
+    rho_k = u[(k-1)*5+4]
+    rho_v1_k = u[(k-1)*5+5]
+    rho_v2_k = u[(k-1)*5+6]
+    rho_v3_k = u[(k-1)*5+7]
+    total_electron_charge += rho_k * equations.charge_to_mass[k]
+    vk1_plus[k] = rho_v1_k * equations.charge_to_mass[k]
+    vk2_plus[k] = rho_v2_k * equations.charge_to_mass[k]
+    vk3_plus[k] = rho_v3_k * equations.charge_to_mass[k]
+  end
+  vk1_plus ./= total_electron_charge
+  vk2_plus ./= total_electron_charge
+  vk3_plus ./= total_electron_charge
+  v1_plus = sum(vk1_plus)
+  v2_plus = sum(vk2_plus)
+  v3_plus = sum(vk3_plus)
```

Comment:
```suggestion
  v1_plus = sum(vk1_plus) / total_electron_charge
  v2_plus = sum(vk2_plus) / total_electron_charge
  v3_plus = sum(vk3_plus) / total_electron_charge
```
I *assume* (but haven't tested) that this is more efficient. It will change the results slightly, though, due to floating point truncation.

Reply:
This might be more efficient to compute `v1_plus`, `v2_plus`, and `v3_plus`, but I also need `vk1_plus`, `vk2_plus`, and `vk3_plus` in the code.